src/vos/vos_obj.c

```diff
@@ -1928,26 +1938,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
 	krec = rbund.rb_krec;
 	umm = vos_obj2umm(oiter->it_obj);
 
-	if (!vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
-				 &oiter->it_punched, &oiter->it_ilog_info))
+	punched = vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
+				      &oiter->it_punched, &oiter->it_ilog_info);
+
+	bmap = krec->kr_bmap;
+
+	if (bmap & KREC_BF_AGG_OPT && !oiter->it_iter.it_for_discard) {
+		if ((bmap & KREC_BF_AGG_NEEDED) == 0) {
+			if (!punched && oiter->it_punched.pr_epc == 0)
```

Comment:
What's this "pr_epc == 0" for? could you explain it a bit?

Reply:
When we punch something, the subtree flags don't get updated.   So we have to scan a tree if it's punched at a higher level.   This covers that case.  If pr_epc is not zero, it means there is an upper level punch.

Reply:
If the key is punched at a higher level, this will be set.  And since we don't scan subtrees and mark them as agg needed, we simply always run aggregation in this case.